### DISCUSSION TITLE
SMT50 soil moisture and temperature sensor

### DISCUSSION THREAD

**Discussion by MariaGardener** [arendst/Tasmota#11545] [Ideas]
Hi,
I have already sucessfully used the SMT50 soil moisture and temperature sensor with Tasmota and analog input.
The SMT50 delivers 0-3V analog voltage signal for soil moisture and temperature.
The conversion formulas from voltage to soil moisture and voltage to temperature are described in the [datasheet](http://www.truebner.de/sites/default/files/SMT50_Flyer_englisch.pdf). It would be very convenient if these formulas would be already configured within Tasmota, e.g. as additonal option in AdcParam. I wonder if someone could implement that or provide a hint how to implement that myself.
Thanks very much
Maria
ps I would like to start a project on garden irrigation

**Comment by sfromis** [comment]
Looks like straight line fitting. You can get Tasmota to do that by configuring the analog pin for range, and using AdcParam 6 for range. You "just" need two points, where you have the analog value and the desired result for each. With that, the range will be calculated for other values fitting the same line. No need for modifying the base Tasmota code for such a simple calculation.

**Comment by MariaGardener** [comment]
Hi,
I generally agree, that it is a rather simple conversion, but I ran into some problems with AdcParam.

For soil moisture I am using this configuration with an EPS32 TTGO board
AdcParam 6, 0 , 3723 , 0 , 50

which means
6: use ADC linear range remapping parameters
0,3723: input range low value, input range high value
(with ESP32 0 to 3.3V means 0 to 4095, so 3723 means 3V which is the maximum sensor output)
0,50: volumetric water content 0 to 50%

Unfortunately the converted values are shown only as integer values  but I would like to have 1 decimal place.

For temperature I am using this configuration with the same EPS32 TTGO board
AdcParam 6, 0 , 1023 , 0 , 33

which means
6: use ADC linear range remapping parameters
0,1023: input range low value, input range high value
(with ESP32 0 to 3.3V means 0 to 4095, so 1023 means about 0,83V which is about 33 degrees accroding to the datasheet)
0,33: temperature 0 to 33 degrees 

Unfortunately the converted values are shown only as integer values  but I would like to have 1 decimal place also.

I did not check the sensor in the freezer so far to see whats happening with temperature above 0 degrees,
but ran into another problem:

I discovered that the ADC of the ESP32 seems to be pretty bad (or the board?)
I used a regulated power supply at the analog input pins and did a few checks, e.g.

0.5V input: ADC value 437
But the ADC value should be 0.5V/3.3V*4095=620

1V input: ADC value 1052
But the ADC value should be 1V/3.3V*4095=1240

2V input: ADC value 2267
But the ADC value should be 2V/3.3V*4095=2481

3V input: ADV value 3752
But the ADC value should be 3V/3.3V*4095=3722

at 3.15 V input the maximum ADC value of 4095 is already achieved.

at 0.12V there is still an ADC value of 0

For soil moisture measurement this may be still acceptable, but for temperature measurement not.
I wonder if other users have made the same experience.
If yes, then I should probably better go to a digital temperature sensor or an external ADC.

**Reply by barbudor** [reply]
Indeed AdcParams returns an integer. So you need a multiplication factor

For exemple I am using the ADC to measure the voltage of a battery and I convert the value to milliVolts instead of volts

**Reply by MariaGardener** [reply]
Yes, that's possible.
I have also the option to convert raw ADC values in my home automation system into temperature, but since the TTGO board has a nice display the plan was to do the conversion and decimal places already on the board. Or I first send the raw ADC values to the server, convert into float with 1 decimal place and send back for displaying. Having everything done on the EPS32 board would be less complicated (it if would work).

**Reply by barbudor** [reply]
You can also do it with rules and arithmetics on variables and mix Range with those
For example assuming you are not using range but ADC Input:
```
on tele-analog0#range do scale1 %value%,0,3723,0,50 endon
on var1#state do backlog publish tele/%topic%/WATER %var1; display %var1% endon
```
That's very rough idea.
You would need to adapt the publication topic and the display

You can also forget about publishing the scaled value and let the backend do the scaling, and use the rule just for the display
Or mix ADC Range and rule to have ANALOG0#Range between 0 and 500 and use the rule for just a `Div1 10` instead of a scale

Multiple options

**Reply by sfromis** [reply]
The ranges set by AdcParam are not "capped", linear values outside of the two input points will still be fitted to the same line, meaning that negative values should be possible. Try the suggestion from arendst about using FreqRes to get decimals.

It should not be a problem if your ESP32 does not give the values you think should match some voltage, the fitting should work as long as you have two points on the line, and the linear fitting gets "close enough".

**Comment by arendst** [comment]
Try latest development version retruning a float allowing decimal resolution control with command ``FreqRes``
